### PR TITLE
Issue #455: Next in line Rollup (rebased)

### LINKED ISSUE #454: Rollup perfromance on high dimensional space
## What went wrong?
I had an issue using Qbeast with 8 dimensions on a relatively large dataset (~600 **GB),** as the jobs never end in one specific task. When I listed the files created, I noticed that they had very different file sizes, ranging from 1.8MB to 220GB. 

```
 ls -hl webpage_embedded_OPQ_quantized_qbeast/8dim_10_bites_100kcs
total 376G
-rw-r--r-- 1 qbst363636 qbst01 220G Oct 27 01:59 0c6a8d1f-52a5-4f8f-8ecc-cc72405c9b7b.parquet
-rw-r--r-- 1 qbst363636 qbst01 1.1G Oct 27 01:15 42a62ec2-4d12-41f3-8798-72158ec3628d.parquet
-rw-r--r-- 1 qbst363636 qbst01 3.6G Oct 27 01:16 48723494-c348-4b88-8e87-ff00f0a3d9ee.parquet
-rw-r--r-- 1 qbst363636 qbst01 2.4G Oct 27 01:16 7ac2246a-a0d9-441b-b10f-69fff3b3dc7b.parquet
-rw-r--r-- 1 qbst363636 qbst01  13G Oct 27 01:17 84f4f027-bfef-495b-9414-ee476db909a9.parquet
-rw-r--r-- 1 qbst363636 qbst01 3.6G Oct 27 01:16 8dd83175-b71a-404a-bca5-5936abc3c188.parquet
-rw-r--r-- 1 qbst363636 qbst01 2.3G Oct 27 01:16 937fd444-7292-4ad8-bf91-4dc93fe6f3af.parquet
-rw-r--r-- 1 qbst363636 qbst01 2.8G Oct 27 01:16 9ba58775-3c24-4b7e-b545-e80e257eb5e7.parquet
-rw-r--r-- 1 qbst363636 qbst01 3.3G Oct 27 01:16 a9ca33f0-0e90-4fc0-bcec-79e60ec6c55a.parquet
-rw-r--r-- 1 qbst363636 qbst01 9.2G Oct 27 01:17 ab02f4b8-c26d-4f34-8e6d-09bcf73a144f.parquet
-rw-r--r-- 1 qbst363636 qbst01 1.9G Oct 27 01:15 b283eb6d-63ad-4930-a083-e0dd6175d3aa.parquet
-rw-r--r-- 1 qbst363636 qbst01 1.8M Oct 27 01:15 b2f6e0ef-6c5a-4992-950a-7f6ae7576cae.parquet
-rw-r--r-- 1 qbst363636 qbst01 1.1G Oct 27 01:16 bd95fd41-7042-422f-a19f-85a6262688ab.parquet
-rw-r--r-- 1 qbst363636 qbst01 1.4G Oct 27 01:15 c01d558f-00bd-45c1-9cb3-adfb5d533a73.parquet
-rw-r--r-- 1 qbst363636 qbst01 1.2G Oct 27 01:15 cc5ef11b-ddac-4b8d-8de4-7296dca21dcf.parquet
-rw-r--r-- 1 qbst363636 qbst01 1.2G Oct 27 01:15 df401968-02ca-4428-8060-1a7286e32506.parquet
-rw-r--r-- 1 qbst363636 qbst01 2.0G Oct 27 01:16 ec897000-a272-428f-8020-5fc362af5e0d.parquet
-rw-r--r-- 1 qbst363636 qbst01 2.5G Oct 27 01:16 f5c70de8-dfca-42cf-8116-9333ff68a2f5.parquet
-rw-r--r-- 1 qbst363636 qbst01 106G Oct 27 01:59 f79b8a20-44a0-4d62-86c9-6feddeda86e5.parquet
```

I suspect this is caused by the roll-up algorithms, which simply push data into the father cube, and, in the case of 8 dimensions, it means a rollup could put together 2^8 + 1= 257 cubes.

## How to reproduce?

Different steps about how to reproduce the problem.

### 1. Code that triggered the bug, or steps to reproduce:

```scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import scala.math._
import io.qbeast.spark.QbeastSparkSessionExtension._
import org.apache.spark.sql.catalyst.TableIdentifier

val spark = SparkSession.builder.appName("SkewedColumnExample").getOrCreate()

// Generate a DataFrame with a range of 100,000 rows
val df = spark.range(1000000)

// Define columns with various distributions
val skewedDF = df
  .withColumn("uniform", rand()) // Uniform distribution [0, 1]
  .withColumn("right_skewed_exp", exp(rand() * 3)) // Exponential distribution (right-skewed)
  .withColumn("left_skewed_log", -log(rand() + 0.01)) // Left-skewed distribution
  .withColumn("normal_dist", randn()) // Normally distributed values
  .withColumn("bimodal_dist", when(rand() < 0.5, rand() * 2).otherwise(rand() * 5 + 5)) // Bimodal distribution
  .withColumn("power_law", pow(rand(), -1.5)) // Power-law distribution (heavy tail)
  .withColumn("triangular_dist", (rand() + rand() + rand()) / 3) // Triangular distribution
  .withColumn("binary_dist", when(rand() > 0.5, 1).otherwise(0)) // Binary distribution (0 or 1)



// Write the DataFrame in Qbeast format with all columns indexed
(skewedDF.write
  .format("qbeast")
  .option("cubeSize",1000)
  .option("columnsToIndex", "uniform,right_skewed_exp,left_skewed_log,normal_dist,bimodal_dist,power_law,triangular_dist,binary_dist")
  .mode("overwrite")
  .saveAsTable("tmp_cesare.dim8test"))


  val location = spark.sessionState.catalog.getTableMetadata(TableIdentifier("dim8test",Some("tmp_cesare"))).location

  val qt = QbeastTable.forPath(spark,location.toString)

  println(qt.getIndexMetrics())

  qt.getDenormalizedBlocks().groupBy("filePath").agg(sum("blockElementCount")).summary().show()

```
And I get this results, where we can see that most of the files have way more elements than the 1000 elements of the desired Cube size (up to 20x). 
```
+-------+--------------------+----------------------+
|summary|            filePath|sum(blockElementCount)|
+-------+--------------------+----------------------+
|  count|                 312|                   312|
|   mean|                NULL|     3205.128205128205|
| stddev|                NULL|     2560.482928268479|
|    min|00f4e8cf-630f-49d...|                    34|
|    25%|                NULL|                  1461|
|    50%|                NULL|                  2152|
|    75%|                NULL|                  4314|
|    max|ff567798-b120-44e...|                 20228|
+-------+--------------------+----------------------+
```
### 3. Spark version:
3.5.3

### 4. Hadoop version:
3.3.4

### 5. How are you running Spark?
Spark Standalone, on 4 large nodes (256 GB RAM, 112 CPU each).

### PR DESCRIPTION
**Pull request by cugni**
Fixes #454.

As the code base has changed considerably in the last months, I'm opening a new PR, as rebasing PR #455 has proved difficult. 

@Jiaweihu08 if you want to work on this issue, you should work on this PR.

### CODE REVIEW

File: core/src/main/scala/io/qbeast/spark/writer/Rollup.scala [hunk 1]
```diff
@@ -78,6 +78,37 @@ private[writer] class Rollup(limit: Double) {
...
+   *    0                     root
+   *    1           c0                   c1
+   *    2     c00         c01       c10         c11
+   * The priority queue will process the cube identifiers in the following order:
+   * c00, c01, c10, c11, c0, c1, root.
+   * c00 -> c01 -> c0, c10 -> c11 -> c1, c0 -> c1 -> root
+   */
+  private object CubeIdOrdering extends Ordering[CubeId] {
+
+    override def compare(x: CubeId, y: CubeId): Int = {
```
Comment (cugni) [inline core/src/main/scala/io/qbeast/spark/writer/Rollup.scala:101]: Can you add a few tests that show how this order is different from the default CubeId ordering?

File: core/src/main/scala/io/qbeast/spark/writer/Rollup.scala [hunk 2]
```diff
@@ -78,6 +78,37 @@ private[writer] class Rollup(limit: Double) {
     }.toMap
   }
 
+  private def areSiblings(cube_a: CubeId, cube_b: CubeId): Boolean = {
+    val sameParent = cube_a.parent == cube_b.parent
+    val differentCube = cube_a != cube_b
```
[the remark this answers is not in the document]
Comment (Jiaweihu08) [reply core/src/main/scala/io/qbeast/spark/writer/Rollup.scala:83]: Just trying to be rigorous

[the remark this answers is not in the document]
Comment (Jiaweihu08) [reply core/src/main/scala/io/qbeast/spark/writer/Rollup.scala]: Fixed

File: core/src/main/scala/io/qbeast/spark/writer/Rollup.scala (hunk 1 above)
Comment (Jiaweihu08) [reply core/src/main/scala/io/qbeast/spark/writer/Rollup.scala:102]: Done

**Comment by Jiaweihu08** [comment]
Here is a simple illustration of how the two rollup implementations differ.

The existing implementation maps a small cube to its parent cube, while the one proposed here first checks the next sibling cube. 

If the number of indexing columns `d,` is three, each inner cube has `2^d = 8` possible child cubes.

The existing implementation, therefore, can theoretically have up to `8 * (cubeSize - 1) + cubeSize ` records in a single file - 8 almost full child cubes, with (cubeSize -1) records all rolled up to their full parent cube.

Conversely, this number is reduced to `2 * (cubeSize - 1) + cubeSize` for the new implementation - a cube `c` is merged with one sibling cube from the left and a child cube.

The example here shows how the old version merges all child cubes with their parent while the new implementation first checks the sibling and outputs a more balanced file size(element count).

```scala
// Indexing three columns
val df = (spark
	.range(85000)
	.toDF("id_1")
	.withColumn("id_2", rand())
	.withColumn("id_3", rand())
)
(df
	.write
	.mode("overwrite")
	.format("qbeast")
	.option("columnsToIndex", "id_1,id_2,id_3")
	.option("cubeSize", "10000")
	.save(path)
)
```

### Old Rollup

| Path | Block Element Counts                                                              | Element Count | Num Blocks |
|------|-----------------------------------------------------------|---------------|--------------|
|file_1 | [9278, 9371, 9483, 9415, 9370, 9300, 9343, 10112, 9328] | 85000              | 9                   |

### New Rollup

| Path | Block Element Counts | Element Count | Num Blocks |
|------|-----------------------|---------------|-------------|
| file_1 | [9365, 9449]              | 18814               | 2                  |
| file_2 | [9305, 9416]              | 18721               | 2                  |
| file_3 | [9324, 9342]              | 18666              | 2                  |
| file_4 | [9409, 9242]              | 18651              | 2                  |
| file_5 | [10148]                        | 10148              | 1                   |

**Comment by Jiaweihu08** [comment]
# File Size Upperbound (in terms of element count)

|          Implementation      | Element Count Upper-bound |
|-------------------------|------------------------------|
| Current Implementation |        (2 ^ d + 1) * cubeSize       |
|             Next-in-line         |             3 * cubeSize                 |

Note: `d` is the number of indexing dimensions.

The upper bound for the existing implementation increases exponentially with the number of indexing dimensions. While the proposed solution has a fixed value.

An important thing to remember is that while `Rollup` imposes strict conditions on when to map a cube to a receiving cube, the cube element counts are approximations during writes. The final element count of the files will depend on the quality of the cube domain size and max weight estimations, so do expect some files to exceed the values presented here. However, during optimization, we compute element counts with 100% certainty.

### Rollup Basics
Both implementations iterate from the bottom up and check the size of the current cube to determine if it should be mapped to a **receiver** cube; a cube `c` is mapped to a **receiver** cube if and only if its (accumulated) element count is `< cubeSize.` The mapping is done regardless of the element count of the **receiver** cube, and the only difference between the two implementations is how the **receiver** cube is selected.

### Existing Rollup Implementation
The **receiver** for the existing implementation is the parent cube. All cubes are either left where they are or mapped to their parent cube.

Given `d` indexing columns, all cubes can have up to `2^d` children. The worst-case scenario happens when all `2^d` child cubes have `cubeSize - 1` elements, and the parent is full. In this case, the parent cube will have accumulated approximately `(2^d + 1)*cubeSize` records. This value grows exponentially with the number of indexing columns.
IMG_B6FDB8423037-1

### Proposed Rollup Implementation (next-in-lin)
Instead of mapping cubes directly to their parent, the proposed implementation merges them with their next sibling cube, and the parent cube will be the receiver if it doesn't have a sibling.

The worst-case scenario for this implementation happens when a cube `c,` which has `cubeSize` elements, receives data from a left sibling cube and a child cube, both with `cubeSize - 1` elements. The accumulated size of `c` is then approximately `3 * cubeSize,` regardless of the number of dimensions `d.`

IMG_B6FDB8423037-2

The accumulated size of `c` cannot be larger than this number because it can only receive data from two other cubes; this is true regardless of the number of possible sibling or child cubes, making the upper-bound independent of `d.` It cannot receive more than `cubeSize - 1` elements from any cube because it doesn't satisfy the rollup condition for merging cubes.

**Comment by fpj** [comment]
The new proposal impacts even lower dimensions, like 2 and 3, which are more common at the moment, but have higher impact with a larger number of dimensions. 

One of the aspects that is not discussed at all, neither in the current approach nor in the new one, is checking the size of the recipient before merging cubes. There seems to be an implicit assumption that having larger files is better than having some small files. I'd think that a lot of small files is bad, but I'd think that a lot of large files, multiple times the cube size, is not our target either. This approach seems to be prone to producing many large files in some scenarios (I'd probably need to work through an example to convince myself of that).

We might want to apply some concept like hysteresis to determine when to merge cubes if we really want to strike a good balance. It depends on the target and we don't seem to have a clear target defined, though.

On a small note, if we do start merging sideways, then roll up makes less of a sense.,

**Comment by cugni** [comment]
Those are very good points. Indeed, we are mixing two concepts here. We use the desiredCubeSize to target the ideal **cube** size, **and** we use the same value for the minimum **file** size during the rollup process. The rationale behind having the _minFileSize_ equal to _desiredCubeSize_  is that in the ideal index state, the one that minimizes the files' column min-max ranges, every (not leaf) cube is stored in a single file alone (with its leaf children).  

As we follow this strict no-files-smaller-than-cube-size approach, we push in the next cube in line even if it already has the desired number of elements. 

We could be using more complex methods like:

1. Merge in the next if I'm more than X%  (e.g., 80%) of the desired cube size. This softens the issue a bit, but it doesn't solve it. This would be the equivalent of having _minFileSize_ = 0.8 * _desiredCubeSize_. While simple, it is not clear to me how to find the proper value for this parameter other than testing and trying different values for each use case. 
2. Find the ideal combination of sibling cubes that minimize the file size difference. This approach can be expensive with high dimensions and might require some heuristics. At the same time, it might lead to worse file-skipping, as the optimal packaging might combine cubes that contain very diverse data (for instance, they have the opposite half of each dimension partition).   With the current approach, we prioritize packaging cubes that have more space in common, using locally a z-order heuristic. 

**TL;DR:** The roll(next?) prioritizes having files bigger than _desiredCubeSize_ and producing better min-max intervals than minimizing the difference between the various file sizes and the _desiredCubeSize_.

**Comment by osopardo1** [comment]
Closing because of unactivity
